### PR TITLE
Run Chrome tests in travis

### PR DESCRIPTION
**Pull request by equalsraf**
Split the travis tests from #9.

Right now this is the same as #9 with a two commits on top to actually test with chrome, just so we can verify #9 works as intended.

TODO: 

- [x] See if we can run tests without sudo, by disabling the sandbox https://github.com/travis-ci/travis-ci/issues/7150#issuecomment-359145343
- [x] Find a way to write driver independent tests - macros seem like an option but I have yet to find one way that does not required nightly
- [x] check if we can remove the addon lines or the package lines in travis https://github.com/fluffysquirrels/webdriver_client_rust/commit/11aedcf2ff8c6894583b8f5a28febcdc1681a484#r27204188
- [x] update README

### CODE REVIEW

**Comment by equalsraf** [comment]
Or maybe something like feature gates for driver specific tests, https://www.reddit.com/r/rust/comments/3i1nki/how_to_skip_expensive_tests_with_cargo_test/ but it does not seem ideal either.

File: src/messages.rs [hunk 1]
```diff
@@ -33,34 +34,44 @@ pub struct WebDriverError {
...
 #[derive(Serialize)]
 pub struct NewSessionCmd {
-    capabilities: JsonValue,
+    capabilities: Capabilities,
 }
 
 impl NewSessionCmd {
     pub fn new() -> Self {
+        let mut capabilities: Capabilities = Default::default();
+        capabilities.alwaysMatch.insert("goog:chromeOptions".to_string(), json!({"w3c": true}));
```
Comment (fluffysquirrels) [inline src/messages.rs]: Can we push `goog:chromeOptions` down into a manual implementation of `Default` on `struct Capabilities`? That seems a little less surprising to me and allows the user to create a session with complete control over capabilities if required.

File: src/messages.rs [hunk 2]
```diff
@@ -33,34 +34,44 @@ pub struct WebDriverError {
...
+}
+
 #[derive(Serialize)]
 pub struct NewSessionCmd {
-    capabilities: JsonValue,
+    capabilities: Capabilities,
 }
 
 impl NewSessionCmd {
     pub fn new() -> Self {
```
Comment (fluffysquirrels) [inline src/messages.rs]: Take a Capabilities value as an argument here, I think.

File: src/messages.rs [hunk 3]
```diff
@@ -33,34 +34,44 @@ pub struct WebDriverError {
...
-                }
-            }),
+            capabilities,
         }
     }
 
 // TODO firefox specifc prefs
 // [moz:firefoxOptions][prefs][name] = value;
 
+    pub fn capability(&mut self, name: &str, value: Option<JsonValue>) {
```
Comment (fluffysquirrels) [inline src/messages.rs]: Can we move this setter to `struct Capabilities` and rename it to `always`, `always_match` or similar? The [capabilities spec](https://w3c.github.io/webdriver/webdriver-spec.html#processing-capabilities) describes `alwaysMatch` and `firstMatch` here, which we may add later.

I really like passing the value as an Option<_> to allow insert or remove with one method. 👍

File: src/lib.rs [hunk 4]
```diff
@@ -84,8 +85,8 @@ pub trait Driver {
     /// The url used to connect to this driver
     fn url(&self) -> &str;
     /// Start a session for this driver
-    fn session(self) -> Result<DriverSession, Error> where Self : Sized + 'static {
-        DriverSession::create_session(Box::new(self))
+    fn session(self, params: &NewSessionCmd) -> Result<DriverSession, Error> where Self : Sized + 'static {
```
Comment (fluffysquirrels) [inline src/lib.rs]: This is a (worthwhile) breaking change, so this release will need to be published as v0.2.0.

**Comment by equalsraf** [comment]
Most tests work, but there is definitely something going on with the **property** method.

```
DEBUG:webdriver_client: GET http://localhost:44183/session/01f04760a3ba5e9c331177972d4f9b74/element/81f43d82-70e4-4950-84ca-60e06fbfaadf/property/href
DEBUG:webdriver_client: result status: 404 Not Found
body: 'unknown command: session/01f04760a3ba5e9c331177972d4f9b74/element/81f43d82-70e4-4950-84ca-60e06fbfaadf/property/href'
thread 'chrome::element_attribute_and_property' panicked at 'Error getting property: JsonDecodeError(ErrorImpl { code: ExpectedSomeValue, line: 1, column: 1 })', libcore/result.rs:916:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
DEBUG:webdriver_client: DELETE http://localhost:44183/session/01f04760a3ba5e9c331177972d4f9b74
DEBUG:webdriver_client: result status: 200 OK
body: '{"value":null}'
```

Apparently this is [not implemented](https://bugs.chromium.org/p/chromedriver/issues/detail?id=1936)

File: src/messages.rs (hunk 2 above)
Comment (equalsraf) [reply src/messages.rs]: If the concern is breaking changes then we can also remove new() and add it at a later time, and just leave default.

**Comment by fluffysquirrels** [comment]
Travis reports some build failures: https://travis-ci.org/fluffysquirrels/webdriver_client_rust/builds/335577419

@illicitonion proposed invoking the tests with `--test-threads=1`, which is now on master. If you merge from master (perhaps no rebase this time :) ) this might improve reliability.

File: src/messages.rs (hunk 3 above)
Comment (equalsraf) [reply src/messages.rs]: Other possibilities here are

- a method to merge capabilities
- a build pattern of some kind, that either returns Self or &mut Self

**Comment by fluffysquirrels** [comment]
> Find a way to write driver independent tests - macros seem like an option but I have yet to find one way that does not required nightly

Which bit requires rust nightly? Doesn't [this](https://stackoverflow.com/a/34666891/94819) work on stable?

**Comment by equalsraf** [comment]
> Which bit requires rust nightly? Doesn't this work on stable?

It runs on stable. But my previous versions depended on concat_idents.

**Comment by fluffysquirrels** [comment]
Ah! Thanks for clarifying.

Perhaps we could have:

```rust
mod firefox {
  browser_tests!(setup_firefox)
}

mod chrome {
  browser_tests!(setup_chrome)
}

fn setup_firefox() {}
fn setup_firefox() {}
```

**Comment by equalsraf** [comment]
That would imply moving the test code inside of the macro definition i.e.

```rust
macro_rules! driver_tests {
    ( $modname:ident, $driver_fn:expr ) => {
        mod $modname {
        #[test]
        fn navigation() {
            let (server, sess) = setup();
            let page1 = server.url("/page1.html");
            sess.go(&page1).expect("Error going to page1");
            assert_eq!(&sess.get_current_url().expect("Error getting url [1]"), &page1, "Wrong URL [1]");

(...all the other tests ...)
        }
    }
}

driver_tests!(gecko, setup_gecko)
driver_tests!(chrome, setup_gecko)
```

It should work, might just be a bit hard to edit tests inside the macro wrapper. Maybe move driver generic tests to different file and do an include!.

**Comment by equalsraf** [comment]
Is https://github.com/fluffysquirrels/webdriver_client_rust/pull/14/commits/fe476263e39f15b74a33e6bbdfcd973a6faab779 similar to what you had in mind for the macros?

**Comment by fluffysquirrels** [comment]
Macros: thanks for working out the definition we need, it's starting to look really good!

I thought about this a lot today and wrote up a draft for you below, with comments to explain my reasoning.

```rust
// I think a trait is a little easier to grok than reading the macro documentation
// to understand what expressions are required. Also easier to extend.
//
// The trait can also be re-used by tests for a single browser (i.e. not in the browser_tests macro)
trait TestBrowser : Debug {
    fn session() -> DriverSession { [...] }
    fn driver() -> Box<Driver> { [...] }
}

#[derive(Default, Debug)]
struct FirefoxTest;
impl TestBrowser for FirefoxTest { [...] }

#[derive(Default, Debug)]
struct ChromeTest;
impl TestBrowser for ChromeTest { [...] }

/// Tests defined in this macro are run once per browser. See the macro invocations below.
macro_rules! browser_tests {
    ($mod_name:ident, $browser_type:ty) => {
        mod $mod_name {
            // Single browser-specific function to grok.
            fn browser() -> Box<TestBrowser> { Box::new($browser_type::default()) }

            // The tests are tightly coupled to the per-browser functions / types, so I think I prefer
            // them right next to each other in one file instead of in a separate included file.
            //
            // "might just be a bit hard to edit tests inside the macro wrapper":
            // I think it's clear enough if you've seen it before.
            // FWIW, this sort of pattern is quite common in rust std
            // e.g. https://github.com/rust-lang/rust/blob/56733bc9f8302409a2b6110f422512923c878154/src/libcore/tests/num/mod.rs#L145-L197
            #[test]
            fn first() -> { [...] }
        }
    };
}

browser_tests!(firefox, FirefoxTest);
browser_tests!(chrome, ChromeTest);
```

**Comment by equalsraf** [comment]
Confirmed, removing firefox from packages and keeping the addon line installs firefox 58. Funny part is we could probably xvfb and run headless firefox (have not tested, but should work since ff 57)

**Comment by equalsraf** [comment]
For later reference on can emulate the missing property method in chrome using javascript and an element ref

```rust
let link = sess.find_element("#link_to_page_2", LocationStrategy::Css)
    .expect("Error finding element");

let href = sess.execute(ExecuteCmd {
    script: "return arguments[0].href;".to_owned(),
    // the browser will convert the json reference to a js reference
    args: vec![link.reference().unwrap()],
}).expect("Error executing script");
exec_json.as_str().unwrap();

```

**Comment by fluffysquirrels** [comment]
I'd love to see headless Firefox. The private project I have using this crate is actually going to be on a headless server.

The xvfb stuff is messier, especially the `sleep 3` that's required in the `.travis.yml` config! I'll add a new ticket to work on headless Firefox.

**Comment by fluffysquirrels** [comment]
I think I'm leaning towards replacing rather than extending capabilities.

I'm working on #18 ("Use Firefox -headless"), and I want to add command line args.

I think this:

```rust
                session_params.extend_always_match(
                    json! {
                        "moz:firefoxOptions": {
                            args: ["-headless"]
                        };
                    });
```

is a little neater for the consumer than:

```rust
                session_params.extend_always_match(
                    "moz:firefoxOptions", json!{
                        args: ["-headless"]
                    });
```

It's fewer characters to type (especially when setting multiple capabilities), and the implementation could be simpler (just use the JSON merge algorithm, as you linked on [this serde_json thread](https://github.com/serde-rs/json/issues/377#issuecomment-341490464)).

Thoughts?

**Comment by fluffysquirrels** [comment]
I've built on this work in branch [ff-headless](https://github.com/fluffysquirrels/webdriver_client_rust/tree/ff-headless) and issue #18.

The [latest Travis build on the branch](https://travis-ci.org/fluffysquirrels/webdriver_client_rust/builds/336963475) just passed, and I'm thinking of merging it soonish.

Happy to hear your comments.

**Comment by equalsraf** [comment]
> Thoughts?

Primarily the implication of this change is that the argument to that function MUST be an object otherwisee it would clobber the `alwaysMatch` attribute turning it into a non object, which would be invalid under the spec.

I've pushed an example 

```rust
    pub fn extend_always_match2(&mut self, new_attrs: JsonValue) -> &mut Self {
        if let JsonValue::Object(map) = new_attrs {
            for (k, v) in map.into_iter() {
                Self::merge_obj(self.capabilities.alwaysMatch
                    .entry(k)
                    .or_insert(JsonValue::Null), v);
            }
        } else {
            // ERROR: input is not an object
        }
        self
    }
```

In this one I just ignore the error though. Otherwise it seems to work.

**Comment by equalsraf** [comment]
Continuing on my previous comment, if extend_always_match takes a value instead of (name, value) pair then should the other methods follow the same convention? Its possible to  Same issue as before with error handling.

```rust
    pub fn always_match(&mut self, capability: JsonValue) -> &mut Self {
        if let JsonValue::Object(b) = capability {
            for (k, v) in b.into_iter() {
                self.capabilities.alwaysMatch.insert(k, v);
            }
        }
        self
    }
```

**Comment by equalsraf** [comment]
Mmm, i cant settle on a merge strategy and naming. I'll remove the extend related commits and leave those for another PR.

**Comment by fluffysquirrels** [comment]
Yes, I eventually realised your way (key and value, not object) was much better for the same reason. I edited my comment but I'm not sure if Github notifies you about edits. Sorry for the confusion.

README changes look good. I'll probably merge all this when I sit down at my computer.

This looks great! It'll be very useful to have Chrome tests running in the CI build.

I feel it would be useful to have examples and/or tests that don't run in headless mode. I find this useful for exploratory testing. I may add a separate issue for those.

**Comment by fluffysquirrels** [comment]
Hey, this turned out really well, thanks for all your work!

I've merged it with my Firefox headless stuff and will push to master then publish to crates.io when I get the build on my branch to pass.

**Comment by fluffysquirrels** [comment]
One small point that I updated in my branch but forgot to call out in your review.

I prefer this:
```rust
let cmd = NewSessionCmd::default();
```
to this:
```rust
let cmd: NewSessionCmd = Default::default();
```

Less typing, and I find it a little more straightforward, coming from an OO background.

**Comment by equalsraf** [comment]
Awesome, I've pushed the NewSessionCmd change anyway. should I close this one?

**Comment by fluffysquirrels** [comment]
Merged.

**Pull request closed without merging**